### DISCUSSION TITLE
How the FrameGraph combine with CommandBuffer with Vulkan Driver

### DISCUSSION THREAD

**Discussion by FuXiii** [google/filament#6103] [Q&A]
When call
```CXX
FRenderer::render(FView const* view)
 |_renderInternal(view);
   |_renderJob(rootArena, const_cast<FView&>(*view));
```
When the engine run the `FRenderer::renderJob(ArenaScope& arena, FView& view)`, it will create a `FrameGraph`
```CXX
//in FRenderer::renderJob(...)
FrameGraph fg(engine.getResourceAllocator());
...
//Add some PassNode
...
fg.compile();
fg.execute(driver);
```
And then add some `PassNode` into the `FrameGraph`, the `Command` will be recorded when run `PassNode::Execute()` into a `CommandBuffeer`.
When run `fg.execute(driver)`, the `FrameGraph` will create and destroy image resources.
```CXX
//in FrameGraph::execute(backend::DriverApi& driver)
...
while (first != activePassNodesEnd) 
{
    for (VirtualResource* resource : node->devirtualize) 
    {
            assert_invariant(resource->first == node);
            resource->devirtualize(resourceAllocator);
    }

    // call execute
    FrameGraphResources resources(*this, *node);
    node->execute(resources, driver);

    // destroy concrete resources
    for (VirtualResource* resource : node->destroy) 
    {
            assert_invariant(resource->last == node);
            resource->destroy(resourceAllocator);
    }
}
...
```
After `FrameGraph::execute(...)` the `CommandBuffer` will be `Submit` to execute when run `FRenderer::endFrame()`.
```CXX
FRenderer::endFrame()
   |_driver.endFrame(mFrameId);//it will call: void VulkanDriver::endFrame(uint32_t frameId)
      |_mContext.commands->flush()//it will submit the CommandBuffer to DeviceQueue, will call vkQueueSubmit
```
Before `CommandBuffer` submitted, the `FrameGraph::execute(....)` will call `resource->destroy(resourceAllocator)` to destroy image resource. so when `CommandBuffer` submitting the resource will become "nullptr" or ”invalid”.

I didn't found how `Filament` fixed this issue in the framework.

I need some help.

**Comment by pixelflinger** [comment]
The destroy call is asynchronous, it is itself enqueued into the backend's command buffer.

**Reply by FuXiii** [reply]
In Vulkan to destroy a `VkImage` just call `vkDestroyImage(...)` on host-domain(CPU), there didn't have some `command` in `CommandBuffer` to recorded for destroy `VkImage`. So how to enqueued into the backend's command buffer?
